File: backend/app/api/favorite.py
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app.core.database import SessionLocal

from app.services.favorite_service import (
    add_favorite,
    get_favorites,
    remove_favorite
)

router = APIRouter(
    prefix="/favorites",
    tags=["Favorites"]
)


def get_db():
    db = SessionLocal()
    try:
        yield db
    finally:
        db.close()


@router.post("/")
def create_favorite(
    user_id: int,
    song_id: int,
    db: Session = Depends(get_db)
):
    return add_favorite(db, user_id, song_id)


@router.get("/")
def all_favorites(
    user_id: int,
    db: Session = Depends(get_db)
):
    return get_favorites(db, user_id)


@router.delete("/")
def delete_favorite(
    user_id: int,
    song_id: int,
    db: Session = Depends(get_db)
):
    return remove_favorite(db, user_id, song_id)

File: frontend/src/services/favorite.js
import api from "./api";

export async function addFavorite(userId, songId) {
    const res = await api.post("/favorites/", null, {
        params: {
            user_id: userId,
            song_id: songId,
        },
    });

    return res.data;
}

export async function getFavorites(userId) {
    const res = await api.get("/favorites/", {
        params: {
            user_id: userId,
        },
    });

    return res.data;
}

export async function removeFavorite(userId, songId) {
    const res = await api.delete("/favorites/", {
        params: {
            user_id: userId,
            song_id: songId,
        },
    });

    return res.data;
}